### PR TITLE
Backoffice to frontend

### PR DESCRIPTION
**Pull request by haogatyp**
UBL work packages AP2 and AP3.

### CODE REVIEW

**Comment by haogatyp** [comment]
Yes, a composer re-run is necessary.

File: Classes/Helper/FormDataReader.php [hunk 1]
```diff
@@ -261,8 +261,11 @@ protected function getNewAndUpdatedFiles()
 
     public function uploadError()
     {
-
-        if ($this->formData['primaryFile'] && $this->formData['primaryFile']['error'] != 0) {
+        if (
+            $this->formData['primaryFile'] &&
+            $this->formData['primaryFile']['error'] != 0 &&
```
Comment (claussni) [inline Classes/Helper/FormDataReader.php]: Magic numbers for primary file errors. What is the meaning?

File: Classes/Helper/Slub.php [hunk 2]
```diff
@@ -91,4 +91,32 @@ public function getSubmitterNotice()
...
+    {
+        $slubInfoNode = $this->getSlubXpath()->query('/slub:info');
+
+        if ($slubInfoNode->length == 1) {
+            $note = $this->slubDom->createElement('slub:note');
+            $note->setAttribute('type', 'private');
+            $note->nodeValue = $noteContent;
+            $slubInfoNode->item(0)->appendChild($note);
+        } else {
+            throw new \Exception('Invalid slubInfo data.');
```
Comment (claussni) [inline Classes/Helper/Slub.php]: Are there more exception cases? Is it only about "invalid data"?

File: Classes/Services/Email/Notifier.php [hunk 3]
```diff
@@ -149,6 +149,55 @@ public function sendIngestNotification(\EWW\Dpf\Domain\Model\Document $document)
...
+                if (empty($body)) {
+                    $body = \TYPO3\CMS\Extbase\Utility\LocalizationUtility::translate('LLL:EXT:dpf/Resources/Private/Language/locallang.xlf:notification.registerDocument.admin.body', 'dpf');
+                    $mailType = 'text/plain';
+                }
+
+                $this->sendMail($clientAdminEmail, $subject, $body, $args, $mailType);
+
+            }
+
+        } catch (\Exception $e) {}
```
Comment (claussni) [inline Classes/Services/Email/Notifier.php]: Empty catch block! Errors should be at least logged for later analysis.

File: Classes/Services/Transfer/DocumentTransferManager.php [hunk 4]
```diff
@@ -288,23 +295,23 @@ public function retrieve($remoteId)
     public function delete($document, $state)
     {
 
-        $document->setTransferStatus(Document::TRANSFER_QUEUED);
-        $this->documentRepository->update($document);
+        //$document->setTransferStatus(Document::TRANSFER_QUEUED);
+        //$this->documentRepository->update($document);
 
             switch ($state) {
```
Comment (claussni) [inline Classes/Services/Transfer/DocumentTransferManager.php]: Case statements are essentially returns. Switch statement can be replaced by multiple if-then.

File: Classes/ViewHelpers/ShowStatusViewHelper.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * LICENSE.txt file that was distributed with this source code.
+ *
+ * The TYPO3 project - inspiring people to share!
+ */
+
+class ShowStatusViewHelper extends \TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper
+{
+
+    /**
+     *
```
Comment (claussni) [inline Classes/ViewHelpers/ShowStatusViewHelper.php]: Maybe the view helper needs a little bit more documentation. ;-)

File: Resources/Private/Language/de.locallang.xlf [hunk 6]
```diff
@@ -495,6 +797,26 @@ Sächsische Landesbibliothek- Staats- und Universitätsbibliothek Dresden
 			<source>You have registered a document</source>
 			<target>Ihre Veröffentlichung auf Qucosa – Anmeldebestätigung</target>
 		</trans-unit>
+		<trans-unit id="notification.registerDocument.admin.body" approved="yes" xml:space="preserve">
```
Comment (claussni) [inline Resources/Private/Language/de.locallang.xlf]: The email message are usually configured in the client record via TYPO3 backend. Is this no longer the case here? If so, why?

File: Resources/Private/Language/locallang.xlf [hunk 7]
```diff
@@ -356,6 +584,18 @@ Date: ###DATE###
 		<trans-unit id="notification.newDocument.submitter.subject">
 			<source>You have registered a document</source>
 		</trans-unit>
+		<trans-unit id="notification.registerDocument.admin.body" xml:space="preserve">
```
Comment (claussni) [inline Resources/Private/Language/locallang.xlf]: Same comment as above. Emails texts are usually configured.

File: Resources/Public/JavaScript/bootstrap.min.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
...
\ No newline at end of file
+  * Bootstrap v4.1.3 (https://getbootstrap.com/)
+  * Copyright 2011-2018 The Bootstrap Authors (https://github.com/twbs/bootstrap/graphs/contributors)
+  * Licensed under MIT (https://github.com/twbs/bootstrap/blob/master/LICENSE)
```
Comment (claussni) [inline Resources/Public/JavaScript/bootstrap.min.js]: @albig MIT licenses are listed as "compatible" with GPL licenses: https://www.gnu.org/licenses/license-list.html#GPLCompatibleLicenses.

> The MIT License is a permissive free software license originating at the Massachusetts Institute of Technology (MIT) in the late 1980s. As a permissive license, it puts only very limited restriction on reuse and has, therefore, reasonable license compatibility.

From https://en.wikipedia.org/wiki/MIT_License

File: ext_emconf.php [hunk 9]
```diff
@@ -27,7 +27,7 @@
     'uploadfolder'     => '1',
     'createDirs'       => 'uploads/tx_dpf',
     'clearCacheOnLoad' => 0,
-    'version'          => '2.0.0',
+    'version'          => '3.0.0',
```
Comment (claussni) [inline ext_emconf.php]: INFO: In future we are going to tag versions from this branch as 4.x

File: Classes/Controller/AbstractDocumentFormController.php [hunk 10]
```diff
@@ -389,6 +367,11 @@ public function initializeAction()
         parent::initializeAction();
     }
 
+    protected function redirectAfterUpdate($document)
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Controller/AbstractDocumentFormController.php]: Parameter was removed.

File: Classes/Controller/SearchController.php [hunk 11]
```diff
@@ -413,4 +484,9 @@ public function searchLatest()
         return $query;
     }
 
+    public function initializeAction()
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Controller/SearchController.php]: Removed, not necessary .

File: Classes/Domain/Model/Document.php [hunk 12]
```diff
@@ -633,160 +657,227 @@ public function setDateIssued($dateIssued)
...
+        return $this->state;
+    }
+
+    public function setState($state)
+    {
+        $this->state = $state;
+    }
+
+    public function getRemoteState() {
+        return explode(':', $this->state)[1];
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Model/Document.php]: Changed code to handle that case.

File: Classes/Domain/Model/DocumentForm.php [hunk 13]
```diff
@@ -274,4 +228,41 @@ public function setProcessNumber($processNumber)
         $this->processNumber = $processNumber;
     }
 
+    /**
+     * Getss the temporary
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Model/DocumentForm.php]: Temporary indicates that a document is a temporary working copy, those documents are used to edit publications on the fly without using a working copy. Changed into isTemporary().

File: Classes/Domain/Model/File.php [hunk 14]
```diff
@@ -345,4 +345,20 @@ public function setDatastreamIdentifier($datastreamIdentifier)
         $this->datastreamIdentifier = $datastreamIdentifier;
     }
 
+    public function copy($fileToCopy) {
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Model/File.php]: File is a model containing metadata of a file related to a document. The function just copies the data of the given file object into the current file object.

File: Classes/Domain/Workflow/DocumentWorkflow.php [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+                self::STATE_IN_PROGRESS_ACTIVE,
+                self::STATE_IN_PROGRESS_INACTIVE,
+                self::STATE_IN_PROGRESS_DELETED,
+                self::STATE_NONE_ACTIVE,
+                self::STATE_NONE_INACTIVE,
+                self::STATE_NONE_DELETED
+            ],
+            "to" => self::STATE_NONE_ACTIVE
+        ],
+/*
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Workflow/DocumentWorkflow.php]: Has been removed.

File: Classes/Helper/Slub.php (hunk 2 above)
Comment (haogatyp) [reply Classes/Helper/Slub.php]: It should be about data. If the node slub:info does not exist the data of the current document is incorrect and we can not add any element to it.

File: Classes/Services/Email/Notifier.php (hunk 3 above)
Comment (haogatyp) [reply Classes/Services/Email/Notifier.php]: The errors are now logged with the typo3 logger.

File: Classes/Services/Transfer/DocumentTransferManager.php (hunk 4 above)
Comment (haogatyp) [reply Classes/Services/Transfer/DocumentTransferManager.php]: The switch statement has been replaced.

File: Classes/ViewHelpers/CanAccessViewHelper.php [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+ * The TYPO3 project - inspiring people to share!
+ */
+
+use TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Object\ObjectManager;
+use EWW\Dpf\Domain\Repository\DocumentRepository;
+use EWW\Dpf\Security\AuthorizationChecker;
+
+class CanAccessViewHelper extends AbstractViewHelper
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/ViewHelpers/CanAccessViewHelper.php]: The Viewhelper has been renamed to IsAccessGrantedViewHelper.

File: Classes/ViewHelpers/ShowStatusViewHelper.php (hunk 5 above)
Comment (haogatyp) [reply Classes/ViewHelpers/ShowStatusViewHelper.php]: Now the view helper returns a user friendly state name to be shown in the document list or the search result list.

File: Resources/Private/Language/de.locallang.xlf (hunk 6 above)
Comment (haogatyp) [reply Resources/Private/Language/de.locallang.xlf]: But default messages have always been configured, this is done in the the same way as before, so it is not mandatory to configure them in the client

File: Resources/Private/Language/locallang.xlf (hunk 7 above)
Comment (haogatyp) [reply Resources/Private/Language/locallang.xlf]: But default messages have always been configured, this is done in the the same way as before, so it is not mandatory to configure them in the client

File: Resources/Public/JavaScript/bootstrap.min.js (hunk 8 above)
Comment (haogatyp) [reply Resources/Public/JavaScript/bootstrap.min.js]: So we do not have a problem here?

File: ext_emconf.php (hunk 9 above)
Comment (haogatyp) [reply ext_emconf.php]: Version has been changed.

File: Classes/Helper/FormDataReader.php (hunk 1 above)
Comment (haogatyp) [reply Classes/Helper/FormDataReader.php]: The numbers have been replaced by php constants for file upload errors.

File: Resources/Public/JavaScript/bootstrap.min.js (hunk 8 above)
Comment (claussni) [reply Resources/Public/JavaScript/bootstrap.min.js]: Seems not. Also embedding MIT licensed code is common in TYPO3.

**Pull request merged**